[macOS] Enable JPEG XL on downlevel OSes

litherum opened the pull request:
#### 6c63375bf01a3f1e853a396adcedf5a10ac740d0
<pre>
[macOS] Enable JPEG XL on downlevel OSes
<a href="https://bugs.webkit.org/show_bug.cgi?id=257768">https://bugs.webkit.org/show_bug.cgi?id=257768</a>
rdar://110349884

Reviewed by Alexey Proskuryakov.

When Apple&apos;s internal build system builds WebKit for downlevel OSes, AppleJPEGXL.framework will
be present, and we can use it. Also, when the installer for new Safari on downlevel OSes runs,
it will install AppleJPEGXL.framework to a place where WebKit can use it.

However, when people build WebKit themselves on downlevel OSes, AppleJPEGXL.framework won&apos;t
be present. But, we want locally-built versions of WebKit to be _able_ to use
AppleJPEGXL.framework if it just happens to be present - we want to be able to test the feature
*before* Apple&apos;s internal build system produces the official build. So we can&apos;t just turn the
feature off entirely on local builds; instead, we have to do something more complicated.

What we want is to create a minimal stub of AppleJPEGXL.framework, check it in to the tree, and
link to it at build time on downlevels using `-weak_framework`. This allows WebKit to use the
framework if it&apos;s present at build time, and not use it if it&apos;s not. Inside WebKit, we can
detect the framework&apos;s presence by declaring one of the external functions with
`__attribute__((weak_import))`, and then just checking to see if the function is null.

Doing it like this means that all downlevel builds will all build with `USE(JPEGXL)`, and will
check at runtime whether or not it can actually open AppleJPEGXL.framework, and gracefully
fall back if it can&apos;t. If we want to test the feature, we just build WebKit, and manually place
AppleJPEGXL.framework at a place where WebKit can see it, and WebKit will automatically
use it. However, if you don&apos;t have a copy of AppleJPEGXL.framework on your development machine,
that&apos;s fine; WebKit will happily run without it.

* Configurations/CommonBase.xcconfig:
* Configurations/WebKitProjectPaths.xcconfig:
* Makefile:
* Source/WebCore/Configurations/WebCore.xcconfig:
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/spi/cocoa/AppleJPEGXL/AppleJPEGXLSPI.h: Added.
* Source/WebCore/PAL/pal/spi/cocoa/AppleJPEGXL/LICENSE.h: Added.
* Source/WebCore/platform/image-decoders/ScalableImageDecoder.cpp:
* Source/WebCore/platform/image-decoders/jpegxl/JPEGXLImageDecoder.cpp:
(WebCore::JPEGXLImageDecoder::create):
* Source/WebCore/platform/image-decoders/jpegxl/JPEGXLImageDecoder.h:
* Source/WebCore/platform/image-decoders/jpegxl/JxlDecoderPtr.h:
* WebKitLibraries/DownlevelFrameworkStubs/120000/AppleJPEGXL.framework/AppleJPEGXL.tbd: Added.
* WebKitLibraries/DownlevelFrameworkStubs/130000/AppleJPEGXL.framework/AppleJPEGXL.tbd: Added.

Canonical link: <a href="https://commits.webkit.org/265050@main">https://commits.webkit.org/265050@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7dc51ace3d56d636d9760d74f7d1b9bcb890383f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/9608 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/9866 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10108 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11269 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9396 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/9619 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/11845 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/9822 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12319 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/9760 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/10615 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8183 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/11428 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/7910 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/8737 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16149 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/8182 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/9005 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/8886 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12225 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/9149 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9375 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/7617 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/9765 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/8566 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/2402 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2311 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/12787 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/10022 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9149 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/2466 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Source/WebCore/Configurations/WebCore.xcconfig:
```diff
@@ -25,7 +25,15 @@
+WK_USE_JPEGXL_PREPROCESSOR_DEFINITIONS = $(WK_USE_JPEGXL_PREPROCESSOR_DEFINITIONS_$(WK_USE_JPEGXL));
+WK_USE_JPEGXL_PREPROCESSOR_DEFINITIONS_YES = USE_JPEGXL=1;
+
+GCC_PREPROCESSOR_DEFINITIONS = $(DEBUG_DEFINES) BUILDING_WEBKIT GL_SILENCE_DEPRECATION=1 GLES_SILENCE_DEPRECATION=1 $(ENABLE_PDFJS) $(ENABLE_LLVM_PROFILE_GENERATION) $(WK_USE_JPEGXL_CONFIGURATION) $(WK_USE_JPEGXL_PREPROCESSOR_DEFINITIONS) $(inherited);
```
What is $(WK_USE_JPEGXL_CONFIGURATION)? Looks like it might be left behind and can be safely deleted.

litherum replied:
Huh. I don't know how that got there. Fixed.

emw-apple commented on Source/WebCore/platform/image-decoders/jpegxl/jxl/decode.h:
```diff
@@ -0,0 +1,32 @@
+
+#pragma once
+
+#if USE(JPEGXL)
```
GTK also has a USE_JPEGXL setting defined in `Source/cmake/OptionsGTK.cmake`. I assume it's nice to have `USE(JPEGXL)` be cross-platform, so maybe a `HAS(ALTERNATE_JPEGXL)` macro that's true on pre-Sonoma macOS is the way to go?

litherum replied:
I think I can just use `USE(JPEGXL) && PLATFORM(COCOA)`. The code that's guarded by this is Cocoa-specific.

I also have to move this header to a `cocoa/` folder so the Linux ports see the real `<jxl/decode.h>` header instead of this one.

aproskuryakov commented on Configurations/CommonBase.xcconfig:
```diff
@@ -25,8 +25,8 @@
 #include "SDKVariant.xcconfig"
-#include "WebKitProjectPaths.xcconfig"
 #include "WebKitTargetConditionals.xcconfig"
+#include "WebKitProjectPaths.xcconfig"
```
Why is this change needed? The includes used to be alphabetically sorted, and someone will eventually sort them again unless there is a comment explaining why that shouldn't be done.

litherum replied:
It’s needed so WK_MACOS_1400 can be used in WebKitProjectPaths.xcconfig. I’ll add a comment.

aproskuryakov replied:
Can `WebKitProjectPaths.xcconfig` include `WebKitTargetConditionals.xcconfig` as a better fix?

aproskuryakov commented on Source/WebCore/Configurations/WebCore.xcconfig:
```diff
@@ -25,7 +25,15 @@
 GCC_PREFIX_HEADER = WebCorePrefix.h;
-GCC_PREPROCESSOR_DEFINITIONS = $(DEBUG_DEFINES) BUILDING_WEBKIT GL_SILENCE_DEPRECATION=1 GLES_SILENCE_DEPRECATION=1 $(ENABLE_PDFJS) $(ENABLE_LLVM_PROFILE_GENERATION) $(inherited);
+
+WK_USE_JPEGXL = $(WK_USE_JPEGXL_$(WK_PLATFORM_NAME));
```
I would name this `WK_USE_APPLEJPEGXL`, because this is about using `AppleJPEGXL.framework`, not about JPEG XL in general.

emw-apple replied:
I think calling it APPLEJPEGXL would also make it easier to notice when we can clean up this logic in a few years.

aproskuryakov commented on Source/WebCore/Configurations/WebCore.xcconfig:
```diff
@@ -25,7 +25,15 @@
+WK_USE_JPEGXL_MACOS_BEFORE_1400 = YES;
+
+WK_USE_JPEGXL_PREPROCESSOR_DEFINITIONS = $(WK_USE_JPEGXL_PREPROCESSOR_DEFINITIONS_$(WK_USE_JPEGXL));
+WK_USE_JPEGXL_PREPROCESSOR_DEFINITIONS_YES = USE_JPEGXL=1;
```
Ditto. I'd name this `WK_APPLEJPEGXL_PREPROCESSOR_DEFINITIONS`

aproskuryakov commented on Configurations/WebKitProjectPaths.xcconfig:
```diff
@@ -77,6 +77,9 @@ WK_PRIVATE_FRAMEWORK_STUBS_DIR_[sdk=iphone*] = $(PROJECT_DIR)/../../WebKitLibrar
 WK_PRIVATE_FRAMEWORK_STUBS_DIR_[sdk=appletv*] = $(PROJECT_DIR)/../../WebKitLibraries/WebKitPrivateFrameworkStubs/appletvos/$(WK_TARGET_TVOS_VERSION_MAJOR);
 WK_PRIVATE_FRAMEWORK_STUBS_DIR_[sdk=watch*] = $(PROJECT_DIR)/../../WebKitLibraries/WebKitPrivateFrameworkStubs/watchos/$(WK_TARGET_WATCHOS_VERSION_MAJOR);
 
+WK_DOWNLEVEL_FRAMEWORK_STUBS_DIR[sdk=macos*] = $(WK_DOWNLEVEL_FRAMEWORK_STUBS_DIR$(WK_MACOS_1400);
```
It's called "downlevel stubs", but it's actually used for all pre-Sonoma builds, including open source ones. Let's just put these in the normal location, `WebKitLibraries/WebKitPrivateFrameworkStubs/Mac`.

litherum replied:
I don’t think that’s right. WebKitPrivateFrameworkStubs Is used for all open source builds, regardless of OS version. This makes sense, as the purpose is to allow open source builds to build with private frameworks.

I’m trying to allow both internal and open source builds to be able to build with a framework that isn’t part of the SDK, but will be present at runtime. This should happen on both internal and open source builds, but only on downlevel OS versions.

Further, I think it’s actually harmful to have internal builds link with WebKitPrivateFrameworkStubs, because for the frameworks in that directory, there should be a version in the internal SDK, and we want to make sure that if the internal framework changes symbols, that we get a build error rather than a runtime error.

So I really do think that WebKitPrivateFrameworkStubs should be a different thing than DownlevelFrameworkStubs.

emw-apple replied:
FWIW, we don't copy the `WebKitLibraries/` directory when running `installsrc` in production builds, so this kind of internal failure could only happen in debug/release Xcode builds.

aproskuryakov replied:
Oh, this means that this PR doesn't fix the problem, at least not entirely. Good point!

litherum replied:
So I need to move DownlevelFrameworkStubs into the Source/ directory somewhere?

emw-apple replied:
It may be easiest to just add `ditto WebKitLibraries/DownlevelFrameworkStubs "$(SRCROOT)/WebKitLibraries/DownlevelFrameworkStubs"`

to the Makefile: https://github.com/WebKit/WebKit/blob/6a68b0c65d6c2504c8455ce9f88b870fb4d6ed94/Makefile#L44-L46.

aproskuryakov replied:
Either that, or change our `Makefile` to perform `installsrc` for more locations. Both changes would be easy to do; I didn't think through it enough to suggest which would be more elegant.

aproskuryakov replied:
(the latter is what Elliott showed how to do, I didn't see that comment when posting mine)

aproskuryakov commented on Source/WebCore/Configurations/WebCore.xcconfig:
```diff
@@ -144,8 +152,11 @@ WK_NO_STATIC_INITIALIZERS_Production_NO_ = $(WK_ERROR_WHEN_LINKING_WITH_STATIC_I
 WK_DEBUG_LDFLAGS =
 WK_DEBUG_LDFLAGS[config=Debug] = -Wl,-debug_variant
 
+WK_LINK_WITH_JPEGXL_LDFLAGS = $(WK_LINK_WITH_JPEGXL_LDFLAGS_$(WK_USE_JPEGXL));
```
Ditto, APPLEJPEGXL.

aproskuryakov commented on Source/WebCore/platform/image-decoders/jpegxl/JxlDecoderPtr.h:
```diff
@@ -28,7 +28,12 @@
 #include <algorithm>
+
+#if PLATFORM(COCOA)
+#include <AppleJPEGXL/decode.h>
```
Is this the right way to include it, not the usual `<AppleJPEGXL/AppleJPEGXL.h>?

litherum replied:
Weirdly, there is no AppleJPEGXL.h file. :/

aproskuryakov reviewed (approved):
r=me with comments.

emw-apple commented on WebKitLibraries/DownlevelFrameworkStubs/120000/AppleJPEGXL.framework/PrivateHeaders/decode.h:
Typically we'd have an SPI header in WebCore that declares the private interfaces it needs, rather than putting a header in the framework stub. That would at least avoid the repetition here.

litherum committed:
[macOS] Enable JPEG XL on downlevel OSes
https://bugs.webkit.org/show_bug.cgi?id=257768
rdar://110349884

Reviewed by Alexey Proskuryakov.

When Apple's internal build system builds WebKit for downlevel OSes, AppleJPEGXL.framework will
be present, and we can use it. Also, when the installer for new Safari on downlevel OSes runs,
it will install AppleJPEGXL.framework to a place where WebKit can use it.

However, when people build WebKit themselves on downlevel OSes, AppleJPEGXL.framework won't
be present. But, we want locally-built versions of WebKit to be _able_ to use
AppleJPEGXL.framework if it just happens to be present - we want to be able to test the feature
*before* Apple's internal build system produces the official build. So we can't just turn the
feature off entirely on local builds; instead, we have to do something more complicated.

What we want is to create a minimal stub of AppleJPEGXL.framework, check it in to the tree, and
link to it at build time on downlevels using `-weak_framework`. This allows WebKit to use the
framework if it's present at build time, and not use it if it's not. Inside WebKit, we can
detect the framework's presence by declaring one of the external functions with
`__attribute__((weak_import))`, and then just checking to see if the function is null.

Doing it like this means that all downlevel builds will all build with `USE(JPEGXL)`, and will
check at runtime whether or not it can actually open AppleJPEGXL.framework, and gracefully
fall back if it can't. If we want to test the feature, we just build WebKit, and manually place
AppleJPEGXL.framework at a place where WebKit can see it, and WebKit will automatically
use it. However, if you don't have a copy of AppleJPEGXL.framework on your development machine,
that's fine; WebKit will happily run without it.

* Configurations/CommonBase.xcconfig:
* Configurations/WebKitProjectPaths.xcconfig:
* Makefile:
* Source/WebCore/Configurations/WebCore.xcconfig:
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/spi/cocoa/AppleJPEGXL/AppleJPEGXLSPI.h: Added.
* Source/WebCore/PAL/pal/spi/cocoa/AppleJPEGXL/LICENSE.h: Added.
* Source/WebCore/platform/image-decoders/ScalableImageDecoder.cpp:
* Source/WebCore/platform/image-decoders/jpegxl/JPEGXLImageDecoder.cpp:
(WebCore::JPEGXLImageDecoder::create):
* Source/WebCore/platform/image-decoders/jpegxl/JPEGXLImageDecoder.h:
* Source/WebCore/platform/image-decoders/jpegxl/JxlDecoderPtr.h:
* WebKitLibraries/DownlevelFrameworkStubs/120000/AppleJPEGXL.framework/AppleJPEGXL.tbd: Added.
* WebKitLibraries/DownlevelFrameworkStubs/130000/AppleJPEGXL.framework/AppleJPEGXL.tbd: Added.

Canonical link: https://commits.webkit.org/265050@main
